applications/FluidDynamicsApplication/custom_utilities/symbolic_stokes_data.h

```diff
@@ -0,0 +1,142 @@
+using ShapeFunctionsType = typename FluidElementData<TDim, TNumNodes, true>::ShapeFunctionsType;
+using MatrixRowType = typename FluidElementData<TDim, TNumNodes, true>::MatrixRowType;
+
+///@}
+///@name Public Members
+///@{
+
+NodalVectorData Velocity;
+NodalVectorData Velocity_OldStep1;
+NodalVectorData Velocity_OldStep2;
+NodalVectorData BodyForce;
+
+NodalScalarData Pressure;
+
+double Density;
+double DynamicViscosity;
+double DeltaTime;      // Time increment
+double DynamicTau;     // Dynamic tau considered in ASGS stabilization coefficients
+
+double bdf0;
+double bdf1;
+double bdf2;
+
+// Auxiliary containers for the symbolically-generated matrices
+BoundedMatrix<double,TNumNodes*(TDim+1),TNumNodes*(TDim+1)> lhs;
+array_1d<double,TNumNodes*(TDim+1)> rhs;
+
+double ElementSize;
+
+///@}
+///@name Public Operations
+///@{
+
+void Initialize(
+    const Element& rElement,
+    const ProcessInfo& rProcessInfo) override
+{
+    // Base class Initialize manages constitutive law parameters
+    FluidElementData<TDim, TNumNodes, true>::Initialize(rElement, rProcessInfo);
+
+    const Geometry< Node<3> >& r_geometry = rElement.GetGeometry();
+    const Properties& r_properties = rElement.GetProperties();
+    this->FillFromNodalData(Velocity, VELOCITY, r_geometry);
+    this->FillFromHistoricalNodalData(Velocity_OldStep1, VELOCITY, r_geometry, 1);
+    this->FillFromHistoricalNodalData(Velocity_OldStep2, VELOCITY, r_geometry, 2);
+    this->FillFromNodalData(BodyForce, BODY_FORCE, r_geometry);
+    this->FillFromNodalData(Pressure, PRESSURE, r_geometry);
+    this->FillFromProperties(Density, DENSITY, r_properties);
+    this->FillFromProperties(DynamicViscosity, DYNAMIC_VISCOSITY, r_properties);
+    this->FillFromProcessInfo(DeltaTime, DELTA_TIME, rProcessInfo);
+    this->FillFromProcessInfo(DynamicTau, DYNAMIC_TAU, rProcessInfo);
+
+    const Vector& BDFVector = rProcessInfo[BDF_COEFFICIENTS];
+    bdf0 = BDFVector[0];
+    bdf1 = BDFVector[1];
+    bdf2 = BDFVector[2];
+
+    ElementSize = ElementSizeCalculator<TDim,TNumNodes>::MinimumElementSize(r_geometry);
+
+    noalias(lhs) = ZeroMatrix(TNumNodes*(TDim+1),TNumNodes*(TDim+1));
+    noalias(rhs) = ZeroVector(TNumNodes*(TDim+1));
+}
+
+void UpdateGeometryValues(
+    const unsigned int IntegrationPointIndex,
+    double NewWeight,
+    const MatrixRowType& rN,
+    const BoundedMatrix<double, TNumNodes, TDim>& rDN_DX) override
+{
+    FluidElementData<TDim,TNumNodes, true>::UpdateGeometryValues(IntegrationPointIndex, NewWeight, rN, rDN_DX);
+}
+
+static int Check(const Element& rElement, const ProcessInfo& rProcessInfo)
+{
+    const Geometry< Node<3> >& r_geometry = rElement.GetGeometry();
+
+    for (unsigned int i = 0; i < TNumNodes; i++)
+    {
+        KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(VELOCITY, r_geometry[i]);
+        KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(BODY_FORCE, r_geometry[i]);
+        KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(PRESSURE, r_geometry[i]);
+    }
+
```

Comment:
Just wondering, why is the check for variables like `DYNAMIC_TAU`, `BDF_COEFFICIENTS` etc. not done here?

Reply:
Because the `DYNAMIC_TAU` and the `BDF_COEFFICIENTS` are stored in the `ProcessInfo` so there is no need to save them as historical nodal variables.